### DISCUSSION TITLE
Module '"@prisma/client"' has no exported member 'PrismaClient'.

### DISCUSSION THREAD

**Discussion by thany** [prisma/prisma#13458] [Q&A]
### Bug description

I'm getting this error in VScode after following this example's [getting started](https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nextjs-api-routes-auth) steps, in the provided `seed.ts` file.

![image](https://user-images.githubusercontent.com/152227/169167299-d199307a-5fbd-4d4f-b591-c33027970696.png)

Other imports work just fine.  
I have never used Prisma before. I'm just following the steps listed.

This problem popped up after the step `npm i`, which tells me either the example is set up wrong, or something specific on my computer is bollocksing up something. No idea. I assume these examples have been tested at some point, so I guess it must be something on my end. Otoh, I have never had this problem with typescript packages if they properly export their things.

### How to reproduce

Literally follow this:
https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nextjs-api-routes-auth

### Expected behavior

No errors.

### Prisma information

I really don't think it's relevant what the schema is. If it is, please have a look at the linked example 😉

### Environment & setup

- OS: Windows 10
- Database: Sqlite
- Node.js version: 16.14.0

### Prisma Version

```
3.14.0
```

**Comment by thany** [comment]
Okay, so the error magically went away the next day (today, at the time of writing this). I didn't change *anything* in the mean time. What gives?

But now the problem is elsewhere. The `PrismaClient` only seems to know about users, accounts, posts, and some other guff. None of that is in my schema. Not anymore. How can I make it know about my schema? I cannot find in the documentation how to deal with this.

So for example, I've put this in the schema:
![image](https://user-images.githubusercontent.com/152227/169403979-f9729b86-884e-4142-8726-974e9c1984c2.png)

And yet:
![image](https://user-images.githubusercontent.com/152227/169404135-06b6f3c1-a6c3-4631-84db-d5bee16b7ade.png)

I must be missing something. It seems as if Prisma is not able to import types from the schema. It's stuck on whatever the project was initialised with.

Update:

But, "migrating" and seeding does somehow manage to work. Making a mistake (again, I have no way to know what a mistake is if the types are completely missing) does throw an error in the CLI. So it knows about types, but doesn't expose them to typescript or something. So I basically have to guess at the types that Prisma should expose, but doesn't, and mentally ignore the potentially tonnes of errors in my project. Doesn't feel right, does it?

Suggestion: about about making schema files just plain typescript? Typescript has more than adequate facilities to transform an interface (which would be a table/model) to its CRUD functions and other supporting types. For instance, styled-components does similar typescript magic - transforming the type of one thing into several supporting types to work with that thing.

**Comment by aladinflux** [comment]
You need to generate the prisma client using `prisma generate` and if you're using VSCode, sometimes, it doesn't pick up the newly generated client.

I find that it helps when you cmd+click into `@prisma/client` so that VSCode parses the file again and updates the types.

**Reply by thany** [reply]
1. Why can't VScode/typescript not pull types directly from the schema without the need for this step?
  Or rather, why is the schema not a typescript file in the first place?

2. More importantly, I guess, I don't see these types being stored anywhere. If they reflect the schema 1-to-1, I feel they should also be checked into source control. But if they don't exist anywhere physically, or if they are stored in some kind of temp/cache folder that obviously is ignored, then they won't be in source control.

**Reply by aladinflux** [reply]
VSCode can't make up types from a schema file. Prisma generates types based on your schema, then VSCode reads them. They are typically stored in node_modules (depending on what package manager you use, usually in `.prisma` folder). You can personalise this behaviour by setting the `output` in your schema (https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#the-location-of-prisma-client).

You can set the output to some folder within your project and commit the output to your source control if you wish to. However, I prefer not to commit those and just `prisma generate` during deployment.

Also, a prisma schema can't be a typscript file (check the generated types to know the amount of work needed), a schema will also generate JS files not just typescript.

**Reply by aladinflux** [reply]
And for VSCode, I think it doesn't quickly pickup the new types due to performance. VSCode can't just keep scanning all the time or your computer will choke. Sometimes you need to help VSCode parse again those types by `cmd+click` into `@prisma/client` and then `cmd+click` into `.prisma` which opens the generated types file and makes VSCode re-read the types.

Another way is to `cmd+shit+P`, then type `Typescript Restart` and select the first option to restart TS server of VSCode which will parse again all types.

**Reply by thany** [reply]
If schema files were typescript, then javascript files (or rather, typescript files that aren't type declaration files) could still be generated.

The amount of types is the *perfect* reason to use typescript, not an excuse to generate it after the fact. It would also solve the problem of how one would derive `UserCreateInput` from the basic `User` interface. Currently there is absolutely no connection between the two.

I would much rather see a `CreateInput<User>`. This also makes it a billion times easier to deal with input and output from generated functions, in generics. This would make it easy, for example, to write a generic function that takes the input of `T` and know upfront what its output has to be.

No sorry mate, but typescript is absolutely _perfect_ for this.

**Reply by aladinflux** [reply]
Thany, how would you define relations?
```
model User {
   somethingId String
   something @relation("SomeRelation", fields: [somethingId], references: [id])
```

How would you define extensions in the schema file?
```
generator { ... }
datasource {.... }
```

How would you add extensions to fields?
```
model Something {
   field String @default("hello") 

  @@map("collectionName")
}
```
How would you differentiate between a model and an embedded document?
```
model User {
  something EmbeddedDoc
}

type EmbeddedDoc {
  someEmbeddedData Int
}
```

How would you define the database type?
```
model Something {
  field1 Int (typescript is just number)
  field2 Float (typescript is number as well)
}
```

Please understand well Prisma before making such comments. The team is making great work and you'll need to understand fully how Prisma works before making such suggestions. Otherwise, just use other node libraries, there are many of them.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Reply by thany** [reply]
Go ahead and close it, but it was not actually answered. Just some talking down on me. This was just some experimentation for me, and so the project is no longer active. I might use it again some day.

**Reply by IamRedzone** [reply]
Crazy how I'm having this problem years later, but seeing how this interaction went, I best go off and figure it myself.

**Reply by tuansheran** [reply]
Same Problem!

**Comment by DmitriyMalayev** [comment]
I had a similar problem and it was resolved by 

rm -rf node_modules
rm -rf .next
npm install
npx prisma generate

**Reply by Sachinvishwa1501** [reply]
Thank You So much it really helped

**Reply by Carlosagdu** [reply]
This really helped

**Reply by saidulislam28** [reply]
this really fixed the problem

**Reply by labeebbappu** [reply]
This works grate for me. still wonder why happend

**Reply by jhonatancaceres** [reply]
Thanks a lot, it worked.

**Reply by MJUrian-Learner** [reply]
Thanks, it worked for me

**Reply by ytpm** [reply]
Did all of that and also closed VSCode / Cursor and the problem solved. Thanks alot!

**Reply by Romeean** [reply]
thank you for help!

**Reply by shamimulhaque1992** [reply]
Thanks for the help, it worked.

**Reply by abbaty48** [reply]
that works out

**Reply by ashongdev** [reply]
This really helped after hours of searching. Thanks

**Reply by prostoleo** [reply]
> I had a similar problem and it was resolved by
> 
> rm -rf node_modules rm -rf .next npm install npx prisma generate

I use Prisma version 7 and import Prisma generated client from `prisma/generated` folder. The error was:

```
The requested module './generated/client.ts' does not provide an export named 'PrismaClient' 
```

I tried this sequence of commands, and still had the same error. The error disappeared after I also removed `prisma/generated` folder. So in my case (Nuxt) it was following order:

1. `rm -rf node_modules`
2. `rm -rf .nuxt`
3. `npm install`
4. `rm -rf prisma/generated` 
5. `npx prisma generate`

**Reply by GusakovskayaDev** [reply]
Personally, neither option suited me. I found a solution by upgrading to the prisma v6.0.0 version.

**Comment by habtesh-beki** [comment]
I had the same problem after I make "npx prisma generate" but you can directly insert the url of PrismaClient folder instead of using "@prisma/client" you can use something "../generated/prisma" like this based on the generated PrismaClient folder

**Reply by loverphp487** [reply]
This really helped

**Reply by mohasinkr** [reply]
This is not a recommended approach right?

**Reply by oallanmendes** [reply]
Nice!

**Reply by coolwolf** [reply]
This cause many errors after running "npm run build" for me.
Then i revert it back to "@prisma/client". This gave not many but single error too.
Then I recheck my whole NextJS folder structure. Tried to organised them in recommended way.
"@prisma/client" still not worked.

Then I revert it back to "import {PrismaClient} from "@/generated/prisma";".
delete node_modules
delete .next
run npm install
run npx prisma generate

And finally "npm run build" worken without any error.

PS: I saw "@/generated/prisma" import is currently recommended way in time I have searched for solution.

**Comment by 3shaan** [comment]
I have faced the same problem. when i commented the output path in schema.prisma file , it generated the output inside the node module and error gone.

**Reply by rgrzesk** [reply]
be aware that in Prisma 7 specifying output will be required: https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/generating-prisma-client

**Reply by 3shaan** [reply]
Thank you. I will keep this in mind.

**Comment by Pankaj-birla** [comment]
Facing similar problems just make the changes in  schema file

 generator client {
  ...
  output = "../../node_modules/.prisma/client"
  ...
}

**Reply by tgiap04** [reply]
this way works fine for me, thanks a lot bro

**Comment by Ritesh1244** [comment]
i was facing same issue ->Module '"@prisma/client"' has no exported member 'PrismaClient' 
I was getting this TypeScript error in my project when trying to import PrismaClient

import { PrismaClient } from '@prisma/client';

but my compiler was throwing error Module '"@prisma/client"' has no exported member 'PrismaClient'. 


so in my case my schema.prisma, I had this generator block:
generator client {
  provider = "prisma-client-js"
  output   = "../generated/prisma"
}

-> so either you have to delete this output   = "../generated/prisma" -> this output part or either you have to 
correctly import in another file like this ->import { PrismaClient } from '../../generated/prisma';

**Reply by 0xdanyal** [reply]
that worked for me and exactly the problem was just wrong path importing like if you are using different output path in schema file than your index (the file where you are exporting it) -- it will give you error
Hence always use correct path based on yours project structure
Mine -- 
import {PrismaClient} from "../generated/prisma"; -- in index.ts where I am exporting
and in shema.prisma -- output   = "../generated/prisma" 

Hence it's just about importing correct path relatively!

Thanks so much!

**Comment by mdranaa** [comment]
**1. Make sure that prisma is installed**
`npm install -D prisma`

**2. Add the generator definition in schema.prisma file**
```
generator client {
  provider = "prisma-client-js"
  output   = "../generated/prisma/client"
}
```

**3. Then install @prisma/client**
`npm install @prisma/client`

**4. Then run the following command**
`npx prisma generate`

**5. Then instantiate prisma client**
`import { PrismaClient } from '../generated/prisma/client'
const prisma = new PrismaClient()
`

This is the best way according to[ documentation](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/generating-prisma-client).

**Reply by MarcusPOliveira** [reply]
Solved for me! Thanks!

**Reply by EoinMoriarty** [reply]
Thanks, worked as well... why was this not working from by default or shown in the docs.

**Reply by Yushank** [reply]
Thanks a lot, you are a saviour. I still don't get it, why some code does import @prisma/client and that works, even I had done that in some of my projects and that worked fine, very strange. 

Anyway thanks man

**Reply by JerateBell** [reply]
Thanks bro

**Reply by VirajMadhu** [reply]
Thanks buddy!! This worked.

**Reply by Humanfrom** [reply]
Thanks, bro! But in my case "output" prop is not nessesary. If you don't specify this property, it will create the client at '@prisma/client', not in a separate folder in the root.

**Reply by WesRamox** [reply]
Thanks bro

**Comment by Hexecoder** [comment]
I opened my schema.prisma file.

I removed the output line inside the generator client block.
(Using a custom output path can sometimes cause this issue.)

Previous example:

```
generator client {
  provider = "prisma-client"
  output   = "../generated/prisma"
}

```
I updated the provider value to prisma-client-js.

Final version:

```

generator client {
  provider = "prisma-client-js"
}

```

Then I ran the following command:

`
npx prisma generate

`
Finally, to clear the TypeScript cache error:

Pressed Ctrl + Shift + P

Selected Restart TS Server

🎯 Result

After these steps, the error was completely resolved.

Hope this helps someone facing the same issue 👍

**Comment by Hexecoder** [comment]
That means:

❌ You can no longer just call new PrismaClient() without configuration.
✅ Prisma 7 now requires an explicit Driver Adapter to connect to Postgres.

### Step 1: Install the Prisma 7 Postgres Adapter 🔌

I opened the terminal and installed the required packages:

```
npm install pg @prisma/adapter-pg
npm install -D @types/pg
```
These packages allow Prisma to connect to Postgres using the new adapter system.

### Step 2: Import the Adapter and Configure Prisma ⚙️

To use the adapter, I imported it in the file where I create the Prisma client (for example, lib/prisma.ts):

```
import { PrismaClient } from "@prisma/client"
import { PrismaPg } from "@prisma/adapter-pg"
import { Pool } from "pg"
```

Then I created a Postgres connection pool and passed the adapter to Prisma:
```
const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
})

const adapter = new PrismaPg(pool)

export const prisma = new PrismaClient({
  adapter,
})
```
The key point here is this:
new PrismaClient() alone is no longer enough in Prisma 7. I must explicitly provide the adapter, otherwise Prisma will not connect to Postgres.

### Step 3: Restart and Test 🚀

I saved the file.

Then I restarted the development server:

`npm run dev
`

Once the server was running again, I went back to Thunder Client and resent the POST request.

This time Prisma will:

- Detect the adapter
- Connect to Supabase
- Return the long-awaited 201 Created response 🎉

**Comment by parsamoloody** [comment]
This is not for not being supported by VsCode, i even tried by using Cursor
read this article for solution: [https://medium.com/@plsreeparvathy/fix-module-prisma-client-has-no-exported-member-prismaclient-16762ce1bdc8](https://medium.com/@plsreeparvathy/fix-module-prisma-client-has-no-exported-member-prismaclient-16762ce1bdc8)

**Comment by parsamoloody** [comment]
if you want to configure your shema.prisma with Default output, it should be like this: 
```
generator client {
  provider = "prisma-client-js"
}
```
then use like this: 
`import { PrismaClient } from '@prisma/client';`

otherwise if you want to configurate with custom, your client configuration should be like this:
```
generator client {
  provider = "prisma-client-js"
  output   = "../generated/prisma"
}
```
then use like this: 
`import { PrismaClient } from '../../prisma/generated';`

**I used the default output for the nestjs app(recommended)**
here is my code:

prisma.service.ts
```
import { Injectable, OnModuleInit, OnModuleDestroy } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';

@Injectable()
export class PrismaService
  extends PrismaClient
  implements OnModuleInit, OnModuleDestroy
{

  async onModuleInit() {
    await this.$connect();
  }

  async onModuleDestroy() {
    await this.$disconnect();
  }
}
```

shema.prisma
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
}
```
